crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1667,22 +1689,82 @@ impl AuthorityPerEpochStore {
+    // non-digest keys.
+    pub async fn notify_read_executed_digests(
+        &self,
+        keys: &[TransactionKey],
+    ) -> SuiResult<Vec<TransactionDigest>> {
+        let non_digest_keys: Vec<_> = keys
+            .iter()
+            .filter_map(|key| {
+                if matches!(key, TransactionKey::Digest(_)) {
+                    None
+                } else {
+                    Some(*key)
+                }
+            })
+            .collect();
+
+        let registrations = self
+            .executed_digests_notify_read
+            .register_all(&non_digest_keys);
+        let executed_digests = self
+            .tables()?
+            .transaction_key_to_digest
+            .multi_get(&non_digest_keys)?;
+        let futures = executed_digests
+            .into_iter()
+            .zip(registrations)
+            .map(|(d, r)| match d {
+                // Note that Some() clause also drops registration that is already fulfilled
+                Some(ready) => Either::Left(futures::future::ready(ready)),
+                None => Either::Right(r),
+            });
+        let mut results = VecDeque::from(join_all(futures).await);
+
+        Ok(keys
+            .iter()
+            .map(|key| {
+                if let TransactionKey::Digest(digest) = key {
+                    *digest
+                } else {
+                    results
+                        .pop_front()
+                        .expect("number of returned results should match number of non-digest keys")
+                }
+            })
+            .collect())
+    }
+
     /// Note: caller usually need to call consensus_message_processed_notify before this call
     pub fn user_signatures_for_checkpoint(
         &self,
+        transactions: &[VerifiedTransaction],
         digests: &[TransactionDigest],
     ) -> SuiResult<Vec<Vec<GenericSignature>>> {
+        assert_eq!(transactions.len(), digests.len());
         let signatures = self
             .tables()?
             .user_signatures_for_checkpoints
             .multi_get(digests)?;
         let mut result = Vec::with_capacity(digests.len());
-        for (signatures, digest) in signatures.into_iter().zip(digests.iter()) {
-            let Some(signatures) = signatures else {
+        for (signatures, transaction) in signatures.into_iter().zip(transactions.iter()) {
+            let signatures = if let Some(signatures) = signatures {
+                signatures
+            } else if matches!(
+                transaction.inner().transaction_data().kind(),
+                TransactionKind::RandomnessStateUpdate(_)
+            ) {
+                // RandomnessStateUpdate transactions don't go through consensus, but
+                // have system-generated signatures that are guaranteed to be the same,
+                // so we can just pull it from the transaction.
+                transaction.tx_signatures().to_vec()
```

Comment:
this is true of all system transactions - would you mind just checking `is_system_transaction()` here instead?